Assignment III
==============

Your ``main.nf`` script should be able to perform ``FASTQC`` and ``MULTIQC`` on your sequencing data. 

I have added a process called ``TX`` - which generates a transcriptome FASTA file for ``Kallisto``.

Update your ``main.nf`` script to include:

1. Transcriptome Indexing. 

2. Kallisto Quantification.

Refer to the `Kallisto documentation <https://pachterlab.github.io/kallisto/manual>`_ and inspect the ``kalisto index`` and ``kallisto quant`` commands. 

Before designing a nextflow workflow, you need to be familiar with the expected outputs generated by the process script body. Shell into your container to run the quantification analysis in bash before implementing the process in nextflow. 

Template script
###############

I have a heavily commented template script for you to use as a starting point: `https://github.com/BarryDigby/rtp_workshop2/blob/dev/template.nf <https://github.com/BarryDigby/rtp_workshop2/blob/dev/template.nf>`_


DAG
###

Hopefully the dag gives you clues how to proceed,.,.,.,.??

.. figure:: /_static/images/flowchart.png
   :figwidth: 700px
   :target: /_static/images/flowchart.png
   :align: center

Advanced script
---------------

As it stands, your ``main.nf`` script will run each process every time we run it. We want to make use of conditional statements to allow users to skip processes - saving computational time and resources.

Refer to the ``Condtionals`` documentation - I have included appropriate parameters and code in both our ``nextflow.config`` and ``main.nf`` scripts for quality control and generating the transcriptome file. 

Your task is to: 

* Include 2 boolean parameters ``kallisto_index`` and ``save_index`` in your ``nextflow.config`` file and script. Use these in a similar fashion to the ``transcriptome`` and ``save_transcriptome`` parameters in the documentation.

We do not need any condional statements for the quantification step - we want to run this every time. 

Please ask for help if anything is unclear.