Assignment III
==============

Your ``main.nf`` script should be able to perform ``FASTQC`` and ``MULTIQC`` on your sequencing data. 

I have added a process called ``TX`` - which generated a transcriptome FASTA file for ``Kallisto``.

Update your ``main.nf`` script to include:

Transcriptome indexing
######################

#. Create a process that creates an index file using the transcriptome fasta file.

   #. Name the process ``INDEX``. 

   #. Include 2 boolean parameters ``kallisto_index`` and ``save_index`` in your ``nextflow.config`` file and script. Use these in a similar fashion to ``transcriptome`` and ``save_transcriptome`` parameters. 

   #. Include a suitable ternary operator after the ``INDEX`` process to accept pre-built index files when supplied to the workflow.

Kallisto quantification
#######################

#. Create a process that performs kallisto quantification using the index file and sequencing reads.

   #. Name the process ``KALLISTO_QUANT``. 

   #. Use the reads staged in ``ch_alignment_reads`` as input to the process - the ``ch_qc_reads`` channel has already been consumed.

|

Refer to the `Kallisto documentation <https://pachterlab.github.io/kallisto/manual>`_ and inspect the ``kalisto index`` and ``kallisto quant`` commands. 

Before designing a nextflow workflow, you need to be familiar with the expected outputs generated by the process script body. Shell into your container to run the quantification analysis in bash before implementing the process in nextflow. 

Template script
###############

I have a heavily commented template script for you to use as a starting point: `https://github.com/BarryDigby/rtp_workshop2/blob/dev/template.nf <https://github.com/BarryDigby/rtp_workshop2/blob/dev/template.nf>`_


DAG
###

Hopefully the dag gives you clues how to proceed,.,.,.,.??

.. figure:: /_static/images/flowchart.png
   :figwidth: 700px
   :target: /_static/images/flowchart.png
   :align: center